### PR TITLE
Feat/158 node angular upgrade  node18 api v9 public admin angular15 using guard

### PR DESCRIPTION
**Pull request by ianliuwk1019**
This PR focus on one API issue after migrating to NestJS V9 (as part of ticket #158):
- 403 always get returned on secured endpoint.
- Previous NestJS custom interceptor for parsing user object not behaving correctly.

The implementation uses NestJS Guards.
Many controllers/endpoints are touched. Some endpoints `auth` from previous version, found to be ambiguous.
I tried my best to interpret that. So, please be careful on this review and we can discuss in details if anything would not be correct. Thanks @basilv .

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-352.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-352.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-352.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-352.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-352.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-352.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

File: api/src/core/security/auth.guard.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    // Deligate validation to `authService.verifyToken(authHeader)`.
+    async validateAuthHeader(request: any, authHeader: string): Promise<boolean> {
+        try {
+            const user = await this.authService.verifyToken(authHeader);
+            this.logger.debug("Authenticated user = %o", user);
+            request.headers['user'] = user;
+            return Promise.resolve(true);
+        }
+        // Need to catch and switch to .resolve(false) for NestJs Guard for correct 403 status.
+        catch (err) {
```
Comment (basilv) [inline api/src/core/security/auth.guard.ts:85]: I'm not sure about catching generic errors and treating as 403, but I see we do the same in AuthService.verifyCognitoToken. By the way, I noticed that verifyCognitoToken calls console.error which is likely not desired (versus calling logger).

File: api/src/core/security/auth.guard.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            this.logger.info("AuthGuard checks failed: %o", err);
```
Comment (basilv) [inline api/src/core/security/auth.guard.ts]: This should likely be at least warn level. If a header is supplied that's invalid, that's either a system error (unlikely) or an attacker, neither seems like should be just 'info' level.

File: api/src/core/security/auth.interceptor.ts [hunk 3]
```diff
@@ -3,6 +3,9 @@ import { PinoLogger } from 'nestjs-pino';
 import { Observable, throwError } from 'rxjs';
 import { AuthService } from './auth.service';
 
+/**
```
Comment (basilv) [inline api/src/core/security/auth.interceptor.ts]: I assume you'll be removing this.

File: api/src/core/security/auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+export const AuthGuardMeta = (...metadata: string[]) => SetMetadata('authGuardMeta', metadata);
+
+// AuthGuard option for scenarios.
+export const GUARD_OPTIONS = {
+    PUBLIC: 'PUBLIC', // Open to public without checks.
+    ANONYMOUS_LIMITED: 'ANONYMOUS_LIMITED', // Mean to be used for 'Open to public' but with 'limited' operational
+            // results, depending on internal business logic. In this case, if `authorization` header is present,
+            // `AuthGuard` will checks; if not present, will 'skip' the check. Endpoint which uses this option needs
+            // to responsible for correct internal business logic for public. 
+    SECURED: 'SECURED' // This case is not needed (Guard validation enforce it by defaul)
```
Comment (basilv) [inline api/src/core/security/auth.guard.ts]: This is the default case if the auth meta is missing, so probably worthwhile to document what actually happens. But I do like the idea of always being explicit with the auth meta (guard option).

Comment (basilv) [reply api/src/core/security/auth.guard.ts]: Hmm, seems like you are using the SECURED option as a default which is fine (e.g. on attachment.controller.ts), so this comment should be changed.

File: api/src/app/modules/forest-client/forest-client.controller.ts [hunk 5]
```diff
@@ -22,14 +23,16 @@ export class ForestClientController extends BaseReadOnlyController<
...
+    @UserHeader() user: User,
   ): Promise<ForestClientResponse[]> {
     return this.service.find(user.clientIds);
   }
 
   @Get(':id')
+  @AuthGuardMeta(GUARD_OPTIONS.PUBLIC)
   @ApiResponse({ status: HttpStatus.OK, type: ForestClientResponse })
   async findOne(@Param('id', ParseIntPipe) id: number): Promise<ForestClientResponse> {
+    console.log("Forest Client findOne: ", id)
```
Comment (basilv) [inline api/src/app/modules/forest-client/forest-client.controller.ts]: Shouldn't use console.log (switch to logger or remove).

**Review by basilv — CHANGES_REQUESTED** [review]
Looks great overall, just some minor comments.

File: api/src/core/security/auth.guard.ts (hunk 1 above)
Comment (ianliuwk1019) [reply api/src/core/security/auth.guard.ts:85]: Good point! I removed the `console.error`.
Also adjust comment on why we catch exception and issue a Promise.resolve(false).
Initially I didn't even try-catch but when testing it I realize NestJS expect only 'boolean/Promise(boolean)' value.  Promise.rejct() won't do and cause NestJS internal 500 error so need to do a try-catch for Promise.resolve(false).
Thank you Basil!

File: api/src/core/security/auth.guard.ts (hunk 2 above)
Comment (ianliuwk1019) [reply api/src/core/security/auth.guard.ts]: Thanks for spotting that. I usually have problem using log level and use it not so correctly.
Changed to warn.

File: api/src/core/security/auth.interceptor.ts (hunk 3 above)
Comment (ianliuwk1019) [reply api/src/core/security/auth.interceptor.ts]: Yes, I will remove it shortly. I leave it there so you have reference when you are on this PR.

File: api/src/app/modules/forest-client/forest-client.controller.ts (hunk 5 above)
Comment (ianliuwk1019) [reply api/src/app/modules/forest-client/forest-client.controller.ts]: Ha, was testing it and forgot to remove it.
Removed now.

File: api/src/core/security/auth.guard.ts (hunk 4 above)
Comment (ianliuwk1019) [reply api/src/core/security/auth.guard.ts]: Thank you Basil. Comment is updated.

**Comment by ianliuwk1019** [comment]
As discussed, merging away this to the parent feature branch.

**Pull request merged**